File: src/go/problems/IsPalindrome.go
// https://leetcode.com/problems/valid-palindrome/

package problems

import (
	"testing"
	"unicode"
)

func TestIsPalindrome(t *testing.T) {
	result := IsPalindrome("A man, a plan, a canal: Panama")
	t.Log(result)
}

func IsPalindrome(s string) bool {
	left := 0
	right := len(s) - 1

	for left < right {
		lRune := rune(s[left])
		rRune := rune(s[right])

		if !(unicode.IsLetter(lRune) || unicode.IsDigit(lRune)) {
			left++
			continue
		}

		if !(unicode.IsLetter(rRune) || unicode.IsDigit(rRune)) {
			right--
			continue
		}

		if unicode.ToLower(lRune) != unicode.ToLower(rRune) {
			return false
		}

		left++
		right--
	}

	return true
}

File: src/csharp/Problems/IsPalindrome.cs
//https://leetcode.com/problems/valid-palindrome/

namespace LeetCode.Problems;

public sealed class IsPalindrome : ProblemBase
{
    [Theory]
    [ClassData(typeof(IsPalindrome))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.Param("A man, a plan, a canal: Panama").Result(true))
          .Add(it => it.Param("A man1, a plan, a canal: Pa1nama").Result(true))
          .Add(it => it.Param("race a car").Result(false))
          .Add(it => it.Param("").Result(true));

    private bool Solution(string s)
    {
        var left = 0;
        var right = s.Length - 1;
        while (left < right)
        {
            if (!char.IsLetterOrDigit(s[left]))
            {
                left++;
                continue;
            }

            if (!char.IsLetterOrDigit(s[right]))
            {
                right--;
                continue;
            }

            if (char.ToLower(s[left]) != char.ToLower(s[right]))
            {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }
}